### PR TITLE
Add rand(seed) function

### PR DESCRIPTION
**Pull request by xerial**
Add a random function that accepts the given seed to produce deterministic random values. For reproducibility of the generates random values, RandCache provides a session-local `Random` object, initialized with the seed value.

A question:
This rand(seed) is not 100% deterministic depending on the number of tasks in a query that executes this function. So should I use `deterministic = false` annotation?

### CODE REVIEW

**Comment by dain** [comment]
How do you expect this to work?  Specifically, what happens with parallel readers on distributed machines?  The engine doesn't provide any guarantees on the order in which rows are processed.

**Comment by electrum** [comment]
This function is definitely non-deterministic, because it is not directly dependent on its inputs (it has state preserved across invocations). So it wouldn't be safe to constant fold it.

Can you explain the motivation for this? I don't see how it is possible to use in any reliable way. This could be useful in something like MySQL where you can guarantee a specific query plan (and they do similar hacky things with variables that you can set as part of a query expression), and anything in Presto that seems to work that way could change in a future release (or based on parallelism, config flags, optimizers, etc.)

**Comment by xerial** [comment]
Stable sampling from a data set is one of the target applications I have in mind. That is useful for preparing training data set for machine learning. Makoto Yui @myui, a developer of Hivemall ( https://github.com/myui/hivemall, machine learning algorithm collections for Hive) may be able to show other usages. 

My expectation is, if the input data (training data) is stable, the same set of splits will be generated. And If we can make the number of tasks stable (e.g., via `min-candidates` parameter), this `rand(seed)` will give the same set of random numbers every time.

**Comment by myui** [comment]
Apache Hive has `rand(seed)` to produce a stable random number for each task.
https://cwiki.apache.org/confluence/display/Hive/LanguageManual+UDF#LanguageManualUDF-MathematicalFunctions

I am assuming the following usage:
```sql
create table src_rnd as
select rand(seed) as rnd, * from src; -- assume src produces same splits

create table train as
select * from src_rnd where rnd < 0.7;

create table test as
select * from src_rnd where rnd >= 0.7;
```

The `create table src_rnd` is idempotent as far as `src` and its splits do not change. The train/test table is reproducible. No matter how each split is allocated to any task executor because each task executor uses a stable seed.

**Comment by dain** [comment]
I don't think the engine can the kind of ordering and only-once-per-row guarantees that this feature would require to work correctly.

If you want a stable random across executions, I can thing of a couple of ways to do this.  
1. Persistent random: When you insert your data, you could add a random column to the data which would always be stable. 
2. Hash row: You could hash the full row data and use that to select.  This is cpu intensive and if you had duplicates, you would get the full set of duplicates.
3. Hash row physical position: Modify hive connector to contain a hidden column for the filename, and row position. Then hash these values.
4. Window random: Write a new window function that generates a random number for each row.  The window operator will establish an ordering and can partition data so you still get parallelism.  For the seed, you would use a hash of the the user supplied seed, and the partition values. 

@cberner might have some more ideas.

**Comment by cberner** [comment]
That's all the ideas I can think of at the moment. (1) seems like the best if this is for one-off cases. If you need to do this a lot then option (3) seems pretty reasonable to me, and it could be useful for other things too.

**Comment by xerial** [comment]
Hash-based option 2 will not yield random values following the uniform distribution, and can be problematic when evaluating the same machine learning algorithm by changing the seed value. 

Having a hidden column (option 3) for the file name, would be useful to generate file name-unique random objects initialized with the given seed. 

Could you suggest how to modify Presto to allow passing such information (e.g., file name) to UDFs? 

Thanks.

**Comment by dain** [comment]
@xerial for option 2, you would initialize the hash function using the seed.  Of course this assumes, you have a good hash function.

For option 3, I would add hidden columns for both the file name and the file row number.  As for how to use a hidden column, they already work in Presto as a normal column.  For example, the TPCH connector has hidden column `row_number` added to every table (https://github.com/facebook/presto/blob/master/presto-tpch/src/main/java/com/facebook/presto/tpch/TpchMetadata.java#L158):
```
select row_number, nationkey, name from tpch.tiny.nation;
 row_number | nationkey |      name      
------------+-----------+----------------
          0 |         0 | ALGERIA        
          1 |         1 | ARGENTINA      
          2 |         2 | BRAZIL         
          3 |         3 | CANADA         
```

**Comment by martint** [comment]
Initializing with a seed in a distributed system is also not guaranteed to produce a good distribution of values. Since all workers will start with the same seed, you'll end up with samples that are not truly statistically independent.

To do so, it'd be necessary to implement something like Java 8's `SplittableRandom`, where the main instance with the seed is created at planning time and each worker is assigned an instance derived from it.

**Comment by xerial** [comment]
@martint 
SplittableRandom approach also looks promising if we can produce stable assignments of the random objects. Maybe we need to use hidden metadata (e.g.,e file name, row number) as @dain suggested.

**Comment by electrum** [comment]
I'm closing this PR since we don't want to add this function. Please feel free to create a new issue and we can discuss concrete alternatives.

**Comment by xerial** [comment]
OK. Thanks for the discussion.
